### ISSUE TITLE
Complete EVSE ID of a charging station

### ISSUE THREAD

**Issue by sideeffect42** [Helium314/SCEE#444]
### General
Affected tag(s) to be modified/added: [ref:EU:EVSE](https://wiki.openstreetmap.org/wiki/Key:ref:EU:EVSE)
Question asked: **Which IDs does this charger have?**

### Checklist
Checklist for quest suggestions (see [guidelines](https://github.com/streetcomplete/StreetComplete/blob/master/QUEST_GUIDELINES.md)):
- [x] 🚧 To be added tag is established and has a useful purpose
       it makes charge points identifiable, possible allowing map apps to query an API if the charger is currently occupied, query charging prices, or detect duplicates should charging stations be imported into OpenStreetMap from an external source.
- [x] 🤔 Any answer the user can give must have an equivalent tagging (Quest should not reappear to other users when solved by one)
- [x] 🐿️ Easily answerable by any pedestrian from the outside but a survey is necessary
      only requirement is to scan a QR code or type the code that's printed below it.
- [x] 💤 Not an overwhelming percentage of quests have the same answer (No spam)
- [x] 🕓 Applies to a reasonable number of map data (Worth the effort)

### Ideas for implementation

**Element selection:** 
The way I understand [amenity:charging_station](https://wiki.openstreetmap.org/wiki/Tag:amenity%3Dcharging_station) this tag is only acceptable for small charging stations where only one charge point is available, so it may make sense to only show this quest on `amenity:charging_station`s which are a node, not an area.

**Metadata needed:** 
No, the codes are unique and in use all over (at least) continental Europe.

**Proposed UI:**

Ideally the UI would display the EVSE IDs already present on the charging station if the number of EVSE IDs is less than either `capacity` or the number of all `socket`s combined.

The user should then have the ability to remove incorrect EVSE IDs or add new ones either by scanning a QR code or typing the ID using the keyboard.

(this suggestion was originally posted as streetcomplete/StreetComplete#5163)

**Comment by HolgerJeromin** [comment]
I think we should skip the "show current ids for keeping the IDs up to date" part for now and ask only for stations without this information.

**Comment by Helium314** [comment]
Adding a QR scanner should be possible, e.g. with [zxing](https://github.com/zxing/zxing) (though it's rather large with ~1 MB).
Is the ID usually present as QR code? Then there should definitely be some way of reading QR codes, either in SCEE or using an external app.
If not, typing the code should be simplified as much as possible, as manually entering numbers/codes is prone to errors.

> I think we should skip the "show current ids for keeping the IDs up to date" part for now and ask only for stations without this information.

That's probably a good idea for a first implementation. Checking/updating IDs can still be implemented later.

**Comment by sideeffect42** [comment]
> Is the ID usually present as QR code? Then there should definitely be some way of reading QR codes, either in SCEE or using an external app.

At least in Switzerland it is common for public charging stations to have a sticker next to the charge ports containing a QR code which redirects you to a website where you can start the charging.

Most commonly it is an Intercharge sticker with a QR code pointing to `http://m.intercharge.eu/qr?evseid=[EVSE-ID]` (e.g. http://m.intercharge.eu/qr/?evseid=CH*CCC*E50585).

(But I believe there are others, too.)

![IMG_20220411_103414](https://github.com/Helium314/SCEE/assets/940027/65ce6a5b-9ecf-4841-8434-c54ea11d6394)

**Comment by Helium314** [comment]
Ideally it would be possible to launch an external QR scanner in a way similar to the StreetMeasure app...
Though if there really are different target URLs of the QR code, extracting the EVSE ID may fail in some cases.

For entering the EVSE ID manually, it should be divided into 3 sub-fields and parts that can be extracted from country and operator should be pre-filled.
E.g. showing 3 input fields like `__*___*_______`, with the first or first 2 parts pre-filled.
QR scan could simply fill the full data, and there should be one such input for each socket.

**Comment by sideeffect42** [comment]
While travelling I paid closer attention to the QR codes located on charging stations and I noticed that while the Intercharge stickers seem to be common in the area I live, this does not apply to other regions as well.

So, autodetecting the EVSE ID from a QR code is probably not as easy as I had originally thought.

Some stations have no QR code at all (but have an EVSE ID in charging apps).
Others have some QR code which either codes the EVSE ID or a part of it directly (e.g. `19XP22T3KK4AD00532*1` for this [charger](https://www.openstreetmap.org/node/11099828119)) or contain an URL with an operator-specific ID (e.g. https://goKWH.me/J4EV for this [charger](https://www.openstreetmap.org/node/9105477014) - it only presents the EVSE ID as a query parameter in the URL when you click "Roaming").

Digitisation, fuck yeah…

**Comment by Helium314** [comment]
Oh, that's "great"...
So the QR scanning would need to consider a lot of different formats, and still (probably) fail when it's just a URL.

But the basic implementation with the text field (or 3 text fields) is still possible.

**Comment by dr2okevin** [comment]
What happened to this? Just having a text-field would be already great.

**Comment by Helium314** [comment]
Nothing happened. That's why it's still open.

**Comment by mnalis** [comment]
> What happened to this? Just having a text-field would be already great.

@dr2okevin Just a note in case you're not aware and you care enough about collecting this data - in SCEE, if you enable `Expert mode`, you will get `Show/edit tags` in `Other answers` for any quest, which allow to manually enter `ref:EU:EVSE` (or any other tag).

Combined with `POI Machine` quest (which will show you all `amenity=charging_station`, even if there are no unanswered quests for them), it already allows you to manually enter it on every charging station, even the ones which have no outstanding quests.

---

_Yes, I know that real separate quest which only shows those without `ref:EU:EVSE` would be preferable, I'm just giving possible alternative until someone decides to write such quest - it is tagged `PR welcome`. For simple text-box it should not be too hard to do, even if one doesn't really have programming experience - by copying and modifying similar quest, see [CONTRIBUTING_A_NEW_QUEST.md](https://github.com/Helium314/SCEE/blob/modified/CONTRIBUTING_A_NEW_QUEST.md). While I'm not interested in the quest itself (I don't have a car, for starters :smiley:), I'm willing to give a hand to someone who is willing to give it a try to write one._

**Comment by Helium314** [comment]
> Combined with `POI Machine` quest (which will show you all `amenity=charging_station`, even if there are no unanswered quests for them)

In 57.0 the things overlay was added, which probably does a better job than the POI quest (but also shows other... things).

> Yes, I know that real separate quest which only shows those without ref:EU:EVSE would be preferable

Here the POI "quest" could be useful again, as it allows adjusting the element filter. You could adjust two of those quests, one showing charging stations with `ref:EU:EVSE` and one without.
(a quest is definitely better here, but more work)

**Comment by mcliquid** [comment]
My idea for a MVP quest scope:
* Target objects: `man_made=charge_point` nodes in the EU **without** `ref:EU:EVSE`
* Question wording: "What is the EVSE ID of this charging point?"
* no QR scanning, no helpers beyond a short example like `DE*ABC*E123456789`
* Tagging result: write the value to `ref:EU:EVSE`
* Behaviour:
  * Quest only appears if the tag is missing
  * User can skip/hide if the ID is not visible. (OR: add `ref:signed=no`)

**Next version:**
* reuse the existing "brewery / destination" quest pattern (simple multi text input) to allow multiple IDs (for `amenity=charging_station`)

**Comment by riQQ** [comment]
A charge point can also have multiple EVSE IDs. See https://wiki.openstreetmap.org/wiki/Tag:man_made%3Dcharge_point.

**Comment by mcliquid** [comment]
That's true. I wanted to keep it simple for the very first version, but one can also use the brewery-quest logic right from the start. Any other blockers?

**Comment by mnalis** [comment]
> * no helpers beyond a short example like `DE*ABC*E123456789`

it would be nice (and hopefully not too hard) to have a basic validator only allowing accepting values matching that  [EVSE syntax](https://wiki.openstreetmap.org/wiki/Key:ref:EU:EVSE#Syntax)
